### ISSUE TITLE
How to create a dataset?

### ISSUE THREAD

**Issue by zuowanbushiwo** [StuartIanNaylor/wake_word_capture#1]
Hi StuartIanNaylor 

How did you create the 'computer' dataset mentioned in your README ? Do you have a complete script?

Your README briefly describes the production process, which is quite different from the process of your https://github.com/StuartIanNaylor/create-dataset and https://github.com/StuartIanNaylor/syn-dataset projects.

Thanks！

**Comment by StuartIanNaylor** [comment]
I decided to do a bigger dataset with more KW and running through it now.
Give me a couple of days and I will update scripts and readme.
I created a new english word lexicon with bigphoney which is great but wow it took over a week :)
'Computer' is just a toy dataset really for proof of concept, the "computer", "felicity", "gregory", "jaqueline","maximus" is just to further extend more classes (equals better) and with the new word lexicon I have class qty's will be upped to 100k+

I doubt because of size and internet connection I will upload that dataset, so those scripts and readme should be updated.
How you get TTS data  and augment is just personal choice, but you can just copy what I do and will document.
Its just more classes and more data equals better classification models aka wakeword.
The sort of urban myth of pouring random into a unknown, wakeword and noise into 2 others with low qty data might be OK as a benchmark, but creates sub par consumer level wakeword.

I am working on it now, have been since the 1st posts, but wow creating my own lexicon that gives syllable qty rather than NLTK/spaCy methods took some considerable time.
Really I should of trained a transformer on cmu_dict and then run through a bigger word list, but not sure if my 8gb Quadro RTX4000 will handle it.
I was hoping my clickbait at https://github.com/idiap/coqui-ai-TTS/discussions/470#discussioncomment-14092682 might spur someone on :)

**Comment by StuartIanNaylor** [comment]
I will leave it open if you have any further questions after complete

**Comment by zuowanbushiwo** [comment]
Hi  StuartIanNaylor

Thank you very much for your reply. I am looking forward to it. 
This is a great learning process for me.  I wonder if I could apply your ideas and techniques to Chinese voice wake-up??

Thanks！

**Comment by StuartIanNaylor** [comment]
Being English means I am terrible with other languages.
Its not so much an idea or technique of mine, just that wake-word models usually have some form of convolution layer that is trying to detect edges and textures as features of usually a MFCC input which is just 'blocky' quantised spectrogram.
So in English the strongest parts of a word are often at the start of each syllable and create edges whilst the textures of MFCC is quite limited.

So rather than just pouring a word soup into a unknown I split it into 2syll, 3yll & 4syll with a 1 sec wakeword of 3syll.
My current list for wakeword is gregory, maximus, josephine, allison, computer all 3syllable words but also phonetically unique start & end. The gre, max, jos, alli & comp and the ory, mus, son, phine & ter should have fairly unique spectra.
Its prob easier doing 3 wakeword than 5, but the more classes of unique features the model uses to identify that class the less likely false positives will be from lack of choice.
All I am doing is trying to group into certain edge types by syllable length, also having some whitespace either side of the wakeword is equally important to the resultant shape in spectra. So don't try to cram a 'Hey Google' in a 1sec model as likely optimal would be just over 1.2 sec maybe as some models have fixed size layers that would benefit from a slight increase. 

Then around each wakeword I create a like class for start and end to find other words that start 'gre' for gregory and end with 'ory', to sort of create a honey pot for false positives and also force the model to train harder to differentiate from the similar wakeword.

Thats it really and its not set in stone and can not make any claim to my method or technique, can you do similar in Chinese?
The only reason its not done is we all seem to lack datasets that are limited when it comes to words , things get poured into a single unknown class because datasets are limited.

So playing about with synthetic data I can input english text but force another language which just helps increase prosidy range as TTS are often very limited in voices and dialect where a language can vary widely in intonation and accent.
I am from the North of England with considerable difference in accent to say london.
So finding TTS emotivoice is great and so are the Sherpa Onnx resources, whilst using Coqui to clone and then force english through 1 of the 14 other languages it supports gives quite a range of prosidy.

https://github.com/Qualcomm-AI-research/bcresnet/tree/main is likely one of the better wakeword models as don't know of another that is so low in parameters, with such high accuracy but the CRNN of https://github.com/google-research/google-research/tree/master/kws_streaming is the one I always test because I know it well.
Can be any model with a dataset that tries to provide class balance without overfitting as most do because the classes are this almost binary afair.

All the better TTS seem to be some sort of transformer model that converts words to tokens that are in a way a phonetic alphabet that the model has created. I am a bit blind to what the tokens are and often use trial and error, to what works.

So much great stuff comes from China opensource, that is far above my level, I am hoping you will be teaching me :)

I will update when I stop making mistakes as being English I do dumb things like use 'felicty' as a wakeword and then realise after almost creating a dataset actually that is 4 syllable and thanks ChatGPT for the list of 3 syllable names I prompted for.

There are some other things that really I am applying too much noise in the dataset augmentation as it should be behind speech enhancement.
For single channel I use https://github.com/breizhn/DTLN that I will retrain with added wakeword data in the dataset, infact likely rehash the dataset as its not that great.
I also find a max9814 microphone and cm108 soundcard off aliexpress better than anything from respeaker or homeassistant but always feed the wakeword with a good AGC which I find https://github.com/radiocicletta/vlevel very good but it does give a bit of latency, but seems much better than the webrtc versions.

**Comment by StuartIanNaylor** [comment]
PS I had a look at your github profile and it would seem you have very similar interests in fixing the initial audio pipeline of speech recognition systems.
The initial analogue gain of the max9814 is far superior to digital gain where upsampling gain of low signals is always a problem.
Combining analogue and digital gain gives this best of both worlds and on my wish list is to up the timing capacitor than the current on sale value to be better optimised for speech and drop the electret for a mems, even if its just cosmetic.
I am a big fan of Radxa and Rockchip and the RK3566 boards Rock3ZeroW & Rock3C are not all that much more $ than a esp32 devkit, a ton of compute with a massively wider range of software.

They will easily run https://github.com/SaneBow/PiDTLN and a wakeword and also have a hardware VAD that I have been failing to work out how to enable, but anyway they make great 'ears'.
I have been going on about this for some time that when speech systems are inherently client/server with ASR/TTS/LLM often residing centrally its pure stupidity to use low compute front end to do final speech enhancement aka HomeAssistant VoicePE.
Be it ESP32 or Arm, distributed 'ears' can have low compute speech enhancement purely to act as broadcast switches to an upstream authoritive wakeword and speech enhancement system as the hardware likely already exists and is running ASR/TTS/LLM. With the diversity of use its rare to get clashes, which can be queued as the system runs faster than realtime.

Be it micro-edge or edge the ears broadcast unprocessed 48khz upstream whilst processing a lower compute 16khz stream purely for wakeword.
Micro-edge this will probably be too much compute but I have wondered about running a snapcast server that will talk to an upstream central client as  https://github.com/yoonsanghyu/FaSNet-TAC-PyTorch is really interesting as it supports "number invariant multi-channel speech separation". I have only ever trained the fixed 2mic version as I am a pytorch noob and not that sure how dynamic inputs work.
The idea is to create number invariant wide array microphone systems using snapcast to sync to 0.2ms and allow users to provide ad-hoc 'ears' of 2 - 6 in a room for source seperation.
This is something I haven't tried as syncing the servers and the 0.2ms of snapcast sync might be detrimental, but worth a try as it would allow really descrete sensors and a hidden central unit doing source seperation.

https://github.com/Rikorose/DeepFilterNet is really excellent but because of [Sonos ](https://github.com/sonos/tract) it just wants a single big core, but is another great upstream filter.

Going back to micro-edge / edge https://k2-fsa.github.io/sherpa/onnx/speech-enhancment/models.html#gtcrn-simple but haven't tested or seen a online / casual example.
Also haven't tested https://github.com/wenet-e2e/wesep
https://google-research.github.io/seanet/basnet/ and https://arxiv.org/pdf/2406.12236v1 and there is code in the wild for binaural speech enhancement totally forgot where its one of the better repos wenet, espnet, sherpa or speechbrain but forgot where doh! I will update and post as it looks quite useful.

Also in the west there are only x2 lowcost dual ADC usb soundcards https://plugable.com/products/usb-audio or https://www.axagon.eu/en/produkty/ada-17 but with x2 Max9814 they make a very good binaral microphone that is plug&play apart from the need for a DiY cable.
I suppose you don't know of a aliexpress like cm108 but 2 channel?

I really like what Wenet did with https://wenet.org.cn/wenet/lm.html and think its disgusting that after it being advocated for use it took a 3 year delay for Homeassistant to ignore and then produce https://github.com/rhasspy/rhasspy-speech which is a refactor and rebrand with no credit to the original! But hey there are a ton of reasons why I will not collab with the BS factory of HA voice, which is a shame as there are a ton of great contributors that are being misled by a rotten core.
Using domain specific Ngram LM is a great idea and nope Speech2Phraise is not original...

I also really like https://github.com/wenet-e2e/wenet/blob/main/runtime/core/websocket/websocket_server.cc and https://k2-fsa.github.io/sherpa/onnx/websocket/index.html
Its really what homeassistant should of done than claiming a totally proprietory socket is an open standard purely to own IP.

However it would be really great to share a common websocket client/server that the server supports multiple clients and the client has possible multiple destination based on upstream busy signals.
I keep banging on about unbranded shared herd of Linux Voice Containers that allow you to string together any chain of speech processes.
That any client will pass realtime websockets audio to any non busy server and when non avail will just queue to file.
Its just a simple addition to either the Sherpa/Wenet websocket offerings that allows splitting and combining process at any point along the audio processing chain that partitions into containers to be free of dependency hell, but also give total control for scalability.
The only other bit that is missing with the 2 packet types of websocket is an accompanying textfile of that stream which is a context history of each processing stage as that is all you need for a complete 'smart voice system' that can be interoperable with any system process.
Well there is another addition and that is to use ALSA file based sources for input audio but you could use a ALSA loopback device (snd-aloop), but https://github.com/voice-engine/alsa_plugin_fifo that got its inspiration from snapcast doing similar is far more flexable.

It would be a major advantage for the likes of wenet, sherpa, espnet & speechbrain to share a common format container that allows users to simply chain offerings from any... LVC only needs a few additions from the great websockets client/servers they already provided.

Still you can use Snapcast to do this as as it does have a great json api that could carry any text, and even allows multichannel audio and compression formats that save network bandwidth, just doesn't fit on micro-edge.

**Comment by zuowanbushiwo** [comment]
Hi StuartIanNaylor

Thank you very much for your detailed reply. I am also planning to study the task of voice wake-up because it is very useful in practice. Currently, the most popular open source OpenWakeWord only supports English.

**Comment by StuartIanNaylor** [comment]
Yeah but I tend to not think of wakeword as any language, purely a wakeup word, but having more natural language wakewords is nice to have.
As said I dunno how Chinese splits into syllables but in the database section you can see the .db that is the source of my words for words that are 'Like KW' by syllable and phonetics and Unknown by syllable.
Create model with multiple KW phonetically focussing on syllable phones and the extra classes create a stronger model, 3 is easier on the brain than 5 :)
I updated my augment script which has got totally messy now as its been hacked so many times it could do with a refactor, works but is untidy.
I added gpuRiR but really for most of us the small furnished rooms tend to have very little reverb.
Also it adds noise but if you are behind a filter noise should be attenuated as RIRs are with time domain source seperation.
If you are going to have a post filter then your dataset should be run through and the augment levels can be higher.
You can have lite fairly innacurate wakeword at the edge that upstream to a authoritive central wakeword even use source seperation and have central wakeword on each seperated stream, but do that upstream.
That means edge wakeword can be just low cost tiny sensors on esp32-s3 and the bc-resnet model is likely a good bet or low cost SBC.

The TTS I use support Chinese so guess its just a matter of feeding with Chinese text than English. Its quite a lot of work to create the dataset and the model but once you do that 'big' one you can use transfer learning and much smaller datasets.
Also the repo is called Wake_Word_Capture as it is easy to do and any model can be fine tuned locally with capture of use, as for the majority of time speech systems are often idle, but again not at the edge but upstream.

Its infrastructure that most get wrong as wakeword is easy, the stream and filter you use at the edge should be purely for the internal wakeword and send unprocessed likely higher bitrate upstream to an authoritive wake that uses higher compute and better speech enhancement filters/source seperation.

Anyway best of luck and check if there is an update as should complete the 5 wakeword model soon. Apols I don't know my "yī, èr, sān" as would other be more help.
